cmd/clusterctl/validation/validate_pods.go

```diff
@@ -0,0 +1,99 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package validation
+
+import (
+	"fmt"
+	"io"
+	"strings"
+
+	"golang.org/x/net/context"
+	corev1 "k8s.io/api/core/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type validationError struct {
+	name    string
+	message string
+}
+
+func ValidatePods(w io.Writer, c client.Client, namespace string) error {
+	fmt.Fprintf(w, "Validating pods in namespace %q\n", namespace)
+
+	pods, err := getPods(c, namespace)
+	if err != nil {
+		return err
+	}
+	return validatePods(w, pods, namespace)
+}
+
+func getPods(c client.Client, namespace string) (*corev1.PodList, error) {
+	pods := &corev1.PodList{}
+	if err := c.List(context.TODO(), client.InNamespace(namespace), pods); err != nil {
+		return nil, fmt.Errorf("Failed to get pods in namespace %q: %v", namespace, err)
+	}
+	return pods, nil
+}
+
+func validatePods(w io.Writer, pods *corev1.PodList, namespace string) error {
```

Comment:
This validates that all pods are running, but it doesn't check for the existence of specific pods that need to be there or for extraneous pods whose health doesn't affect the health of the cluster.

Reply:
@roberthbailey can you elaborate more on how to approach this? 
* I think this is not straight forward to implement, since we need to take number of nodes/masters into account which will impact number of apiserver, kubeproxy, etc. to validate. Also, need to consider internal vs external etcd.

* How do we check for existence of specific pods which are required? Do we check the labels? Are label values standard across k8s versions?

* What are considered required pods? I was thinking 6 types: apiserver, controller-manager, scheduler,  etcd (if internal), kube-proxy, kubedns/coredns. CNI is hard to validate. Also, how do you want to output extraneous pods which are failing, output as warning?

Reply:
I agree with your points, but it doesn't make the current code make more sense to me. 

To expand on your first point, depending on how the cluster is structured, it isn't always possible to see the control plane pods (e.g. if they are running as static pods and the kubelet on the master isn't registered into the cluster). So you may not always be able to check for the apiserver, scheduler, etc. Also, it's possible to create a fully conformant cluster without running kube-proxy on nodes (if you have a different mechanism for doing service forwarding). 

Using pod names or labels doesn't seem like it's going to be a reliable way to find the pods we care about. In fact, "the pods we care about" is a provider-specific notion and while there may be some commonality, I'm not sure how much we can count on.

Reply:
I have added component validation outside of validatePods function if this is acceptable https://github.com/kubernetes-sigs/cluster-api/pull/679/commits/2117bc7df65b7ca366ef679ddb5ea4a6d7e654f3